assets/scripts/info_bubble/BuildingHeightControl.jsx

```diff
@@ -0,0 +1,270 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { connect } from 'react-redux'
+import { t } from '../app/locale'
+import { MAX_BUILDING_HEIGHT, isFlooredBuilding } from '../segments/buildings'
+import { addBuildingFloor, removeBuildingFloor, setBuildingFloorValue } from '../store/actions/street'
+import { _prettifyHeight } from './info_bubble'
+import { KEYS } from '../app/keyboard_commands'
+import { loseAnyFocus } from '../util/focus'
+
+const WIDTH_EDIT_INPUT_DELAY = 200
+
+class BuildingHeightControl extends React.Component {
+  static propTypes = {
+    touch: PropTypes.bool,
+    position: PropTypes.oneOf(['left', 'right']),
+    variant: PropTypes.string,
+    value: PropTypes.number,
+    addBuildingFloor: PropTypes.func,
+    removeBuildingFloor: PropTypes.func,
+    setBuildingFloorValue: PropTypes.func
+  }
+
+  constructor (props) {
+    super(props)
+
+    this.oldValue = null
+    this.timerId = -1
+    this.inputEl = null
+
+    this.state = {
+      isEditing: false,
+      displayValue: _prettifyHeight(props.value)
+    }
+  }
+
+  /**
+   * If UI is not in user-editing mode, update the display
+   * when the value in store changes
+   *
+   * @param {Object} nextProps
+   */
+  componentWillReceiveProps (nextProps) {
+    if (!this.state.isEditing) {
+      this.setState({
+        displayValue: _prettifyHeight(nextProps.value)
+      })
+    }
+  }
+
+  /**
+   * If UI is going to enter user-editing mode, immediately
+   * save the previous value in case editing is cancelled
+   *
+   * @param {Object} nextProps
+   * @param {Object} nextState
+   */
+  componentWillUpdate (nextProps, nextState) {
+    if (!this.state.isEditing && nextState.isEditing) {
+      this.oldValue = this.props.value
+    }
+  }
+
+  onClickIncrement = () => {
+    this.props.addBuildingFloor(this.props.position)
+  }
+
+  onClickDecrement = () => {
+    this.props.removeBuildingFloor(this.props.position)
+  }
+
+  onInput = (event) => {
+    window.clearTimeout(this.timerId)
+
+    const value = event.target.value
+
+    // Update the input element to display user input
+    this.setState({
+      isEditing: true,
+      displayValue: value
+    })
+
+    // However, there is a short delay between inputs and updating the model
+    // to prevent rapid key entry from thrashing things
+    const update = () => {
+      this.props.setBuildingFloorValue(this.props.position, value)
+    }
+
+    this.timerId = window.setTimeout(update, WIDTH_EDIT_INPUT_DELAY)
```

Comment:
Have you thought about using a library to handle debouncing like this? https://www.npmjs.com/package/debounce I'd bet that we need it elsewhere too

Reply:
That's a good idea. What do you think about Lodash's `debounce`? It's fundamentally the same implementation as the debounce package you linked to (code: https://github.com/lodash/lodash/blob/master/debounce.js) and we already include lodash in Streetmix's dependencies.